addons/sourcemod/scripting/VIP_FreeVIPGiveaway.sp

```diff
@@ -1,170 +1,349 @@
+				if(VIP_IsClientVIP(i) && VIP_GetClientID(i) == -1)
+					VIP_RemoveClientVIP2(_, i, false, false);
+
+				continue;
+			} 
+
+			if(VIP_IsClientVIP(i))
+				continue;
+
+			VIP_GiveClientVIP(_, i, 0, vipGroup, false);
 		}
 
 		// push chat message
-		CPrintToChatAll("[SM] {default}Free {pink}VIP {default}Giveaway is {green}enabled{default}. Active players got Free {pink}VIP");
+		CPrintToChatAll("[SM] {pink}Free VIP Giveaway {default}is {green}enabled{default}. {pink}Active players got Free VIP.");
 	}
 	else
 	{
 		char testVipGroup[16];
-		GetConVarString(g_Cvar_TestVIPGroup, testVipGroup, sizeof(testVipGroup));
+		g_Cvar_TestVIPGroup.GetString(testVipGroup, sizeof(testVipGroup));
 
 		// remove temporary vip from players if not enough players
 		for(int i = 1; i <= MaxClients; i++)
 		{
-			if(IsClientInGame(i) && !IsFakeClient(i) && VIP_IsClientVIP(i) && VIP_GetClientID(i) == -1 && !VIP_GetClientVIPGroup(i, testVipGroup, sizeof(testVipGroup)))
-			{
-				VIP_RemoveClientVIP2(_, i, false, false);
-			}
+			if(!IsClientInGame(i) || IsFakeClient(i) || !VIP_IsClientVIP(i) || VIP_GetClientID(i) != -1 || VIP_GetClientVIPGroup(i, testVipGroup, sizeof(testVipGroup)))
+				continue;
+
+			VIP_RemoveClientVIP2(_, i, false, false);
 		}
 
 		int playersNeeded = minPlayers - playersOnServer;
 
 		// push chat message
-		CPrintToChatAll("[SM] {default}Free {pink}VIP {default}Giveaway is {red}disabled{default}.\nPlayers on: {green}%d {default}| Players required: {green}%d {default}| Players needed: {green}+%d", playersOnServer, minPlayers, playersNeeded);
+		CPrintToChatAll("{green}[SM] {pink}Free VIP Giveaway {default}is {red}disabled{default}.\nPlayers on: {green}%d {default}| Players required: {green}%d {default}| Players needed: {green}+%d", playersOnServer, minPlayers, playersNeeded);
 	}
 }
 
-public void OnClientConnected(int client)
+public void VIP_OnClientLoaded(int client)
 {
-	if (GetConVarInt(g_Cvar_MinPlayers) == 0)
+	if(!IsFreeVIPOn())
 	{
-		char vipGroup[16];
-		GetConVarString(g_Cvar_VIPGroup, vipGroup, sizeof(vipGroup));
+		char cookieValue[6];
+		g_hCookie.Get(client, cookieValue, sizeof(cookieValue));
+		/* Set cookie to false in all clients when vip ends */
+		if(StrEqual(cookieValue, "true"))
+			g_hCookie.Set(client, "false");
+			
+		return;
+	}
+		
+	if(IsClientSourceTV(client) || IsFakeClient(client))
+		return;
+		
+	if(!IsClientAuthorized(client))
+		return;
+	
+	GiveVIP(client);
+}
 
-		if (client && !IsFakeClient(client) && !VIP_IsClientVIP(client))
-		{
-			VIP_GiveClientVIP(_, client, 0, vipGroup, false);
-		}
+void GiveVIP(int client)
+{
+	if(!IsFreeVIPOn())
+		return;
+		
+	if(client < 1)
+		return;
+
+	if(VIP_IsClientVIP(client))
+	{
+		SetVIP(client);
+		return;
 	}
+		
+	char vipGroup[16];
+	g_Cvar_VIPGroup.GetString(vipGroup, sizeof(vipGroup));
+	VIP_GiveClientVIP(_, client, 0, vipGroup, false);
+	/* Give client the vip flags */
+	int flags = GetUserFlagBits(client);
+	if(!(flags & ADMFLAG_CUSTOM1))

```

Comment:
dont handle flags here, its VIP_SourcemodFlags that should handle this

Reply:
i will open an issue to do it later

Reply:
I'm reopening a PR. I let you handle it like you said.